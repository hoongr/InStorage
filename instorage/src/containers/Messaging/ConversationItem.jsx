import React from 'react';
import { ListGroup, Button } from 'react-bootstrap'
import Truncate from 'react-truncate'
import { timezone } from '@google/maps/lib/apis/timezone';
import { withRouter } from "react-router-dom";

class ConversationItem extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return(
      <div>
        <style type="text/css">
          {`
            .img-size {
              width: 50%;
              height: auto;
            }
          `}
        </style>
        <ListGroup className="m-auto d-table h150px w90 list-group-horizontal">
          <ListGroup.Item className="d-table-cell align-middle w20 text-center no-border-right">
            <div>
              <img className="circle-profile-pic cropped-pic" src={this.props.profile} alt="me"/>
              {/* <div className='tiny-text'>{this.props.label}</div> */}
            </div>
          </ListGroup.Item>
          <ListGroup.Item className="d-table-cell align-middle w60 no-border-left no-border-right">
            <h4>
              {this.props.label}
            </h4>
          </ListGroup.Item>
          {/* <ListGroup.Item className="d-table-cell align-middle w60 no-border-left no-border-right">
            <div>
              <Truncate lines={3}>
                {this.props.body}
              </Truncate>
            </div>
          </ListGroup.Item> */}
          <ListGroup.Item className="d-table-cell text-center align-middle w30 no-border-left">
            <div>
              <Button size="md" variant="outline-primary"
                onClick={() => {
                  this.props.history.push('/chat/' + this.props.userB)
                }}
              >View messages</Button>
            </div>
          </ListGroup.Item>
        </ListGroup>
      </div>
    );
  }
}

export default withRouter(ConversationItem);