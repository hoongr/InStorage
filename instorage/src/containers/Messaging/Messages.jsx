import React from 'react';
import { ChatFeed, ChatBubble, BubbleGroup, Message } from 'react-chat-ui';
import { withAuthorization } from '../../Session';
import FirebaseManager from '../../api/routes/firebaseManager.js'
const firebase = require("firebase");
// Required for side-effects
require("firebase/firestore");

const users = {
  0: 'You',
  1: 'Evan',
};

const customBubble = props => (
  <div>
    <p>{`${props.message.senderName} ${props.message.id ? 'says' : 'said'}: ${
      props.message.message
    }`}</p>
  </div>
);

class Messages extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      messages: [
        // new Message({
        //   id: 1,
        //   message: 'Hey! Evan here.',
        //   // senderName: 'Evan',
        // }),
      ],
      useCustomBubble: false,
      curr_user: 0,
    };
    this.loadMessages();
  }

  onPress(user) {
    this.setState({ curr_user: user });
  }

  onMessageSubmit(e) {
    const input = this.message;
    e.preventDefault();
    if (!input.value) {
      return false;
    }
    this.pushMessage(this.state.curr_user, input.value);
    input.value = '';
    return true;
  }

  isThisMe(uid) {
    console.log("uid " + uid);
    console.log("curr id " + firebase.auth().currentUser.uid);
    if(uid == firebase.auth().currentUser.uid) {
      console.log("this message is from me");
      return 0;
    } else {
      console.log("this message is not from me");
      return 1;
    }
  }

  loadMessages() {
    console.log('loading');
    // Create the query to load the messages and listen for new ones.
    // const queryDocumentSnapshot = null;
    console.log([firebase.auth().currentUser.uid, this.props.match.params.userB]);
    console.log([firebase.auth().currentUser.uid + '_' + this.props.match.params.userB, this.props.match.params.userB + '_' + firebase.auth().currentUser.uid]);
    firebase.firestore()
                    .collection('chatRooms')
                    .where('members', 'array-contains-any', [firebase.auth().currentUser.uid + '_' + this.props.match.params.userB, this.props.match.params.userB + '_' + firebase.auth().currentUser.uid])
                    .orderBy("timestamp", "desc")
                    .limit(1)
                    .get()
                    .then(querySnapshot => {
                      if (!querySnapshot.empty) {
                          //We know there is one doc in the querySnapshot
                          console.log('chatRoom found!');
                          const queryDocumentSnapshot = querySnapshot.docs[0];
                          const chatRoomId = queryDocumentSnapshot.id;
                          this.setState({chatRoomId: chatRoomId});
                          const _this = this;
                          firebase.firestore()
                            .collection('chatRooms')
                            .doc(chatRoomId)
                            .collection('messages')
                            .orderBy('timestamp', 'asc')
                            .onSnapshot(function(snapshot) {
                            snapshot.docChanges().forEach(function(change) {
                              console.log('snapshot!')
                              if (change.type === 'removed') {
                                console.log('trying to delete message!')
                                // deleteMessage(change.doc.id);
                              } else if(change.type === 'added') {
                                var message = change.doc.data();
                                console.log(change.type);
                                console.log(message);
                                _this.displayMessage(_this.isThisMe(message.uid), message.message);
                              }
                            });
                          });
                      } else {
                          console.log("No chatRoom found, creating new chatRoom...");
                          FirebaseManager.getUserProfile(this.props.match.params.userB)
                          .then(res => {
                            if(res == null){
                              console.log('cant find user');
                              return;
                            }else{
                              firebase.firestore().collection('chatRooms').add({
                                roomName: res._name,
                                members: [firebase.auth().currentUser.uid + '_' + this.props.match.params.userB, this.props.match.params.userB + '_' + firebase.auth().currentUser.uid],
                                participants: [firebase.auth().currentUser.uid, this.props.match.params.userB],
                                timestamp: firebase.firestore.FieldValue.serverTimestamp()
                              }).catch(function(error) {
                                console.error('Error writing new message to database', error);
                              });
                              this.loadMessages();
                            };
                          });
                      }
                    });
    
  }

  displayMessage(recipient, message){
    console.log("displayMessage(): " + recipient + " " + message);
    const prevState = this.state;
    const newMessage = new Message({
      id: recipient,
      message,
      // senderName: users[recipient],
    });
    prevState.messages.push(newMessage);
    this.setState(prevState);
  }

  pushMessage(recipient, message) {
    // Add a new message entry to the database.
    firebase.firestore().collection('chatRooms').doc(this.state.chatRoomId).collection('messages').add({ 
      uid: firebase.auth().currentUser.uid,
      // senderName: users[recipient],
      message: message,
      timestamp: firebase.firestore.FieldValue.serverTimestamp()
    }).catch(function(error) {
      console.error('Error writing new message to database', error);
    });
    console.log("pushMessage(): " + recipient + " " + message);
  }

  render() {
    // if(firebase.auth().currentUser != null){
    //   console.log("current user uid: " + firebase.auth().currentUser.uid);
    // }
    return (
      <div className="vh90">
        <div className="p20 h100 flex_col justify-between chatfeed-wrapper">
          <ChatFeed
            chatBubble={this.state.useCustomBubble && customBubble}
            messages={this.state.messages} // Boolean: list of message objects
            showSenderName
          />

          <form onSubmit={e => this.onMessageSubmit(e)}>
            <input
              ref={m => {
                this.message = m;
              }}
              placeholder="Type a message, then press Enter..."
              className="message-input w100"
            />
          </form>
        </div>
        {/* firebase SDK */}
        <script src="/__/firebase/6.4.0/firebase-app.js"></script>
        <script src="/__/firebase/6.4.0/firebase-auth.js"></script>
        <script src="/__/firebase/6.4.0/firebase-storage.js"></script>
        <script src="/__/firebase/6.4.0/firebase-messaging.js"></script>
        <script src="/__/firebase/6.4.0/firebase-firestore.js"></script>
        <script src="/__/firebase/6.4.0/firebase-performance.js"></script>
      </div>
    );
  }
}

const condition = authUser => !!authUser;
export default withAuthorization(condition)(Messages);
