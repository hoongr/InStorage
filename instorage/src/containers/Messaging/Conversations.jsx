import React from 'react';
import ConversationItem from './ConversationItem';
import { withAuthorization } from '../../Session';
import FirebaseManager from '../../api/routes/firebaseManager.js'
import { withRouter } from "react-router-dom";
const firebase = require("firebase");
require("firebase/firestore");


class Conversations extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = {
      conversations: []
    }
  }

  componentDidMount(){
    const _this = this;
    const { auth } = this.props.firebase;
    var uid = auth.currentUser.uid;
    console.log('user uid ' + uid);
    
    firebase.firestore()
      .collection('chatRooms')
      .where('participants', 'array-contains', uid)
      .get()
      .then(function(querySnapshot) {
        querySnapshot.forEach(function(doc) {
          const data = doc.data();
          const userB = (uid == data.participants[0]) ? data.participants[1] : data.participants[0];
          console.log('userB ' + userB);
          FirebaseManager.getUserProfile(userB)
          .then(res => {
            console.log(res);
            console.log('members ' + data.members);
            // console.log('roomName ' + data.roomName);
            const prevState = _this.state;
            if(typeof res._profilePicURL === 'undefined') {
              console.log('undefined pp');
              prevState.conversations.push({id: doc.id, label: res._name, userB: userB, profile: 'https://cutecatsinhats-x7v0etsjgzjvirs3.netdna-ssl.com/wp-content/uploads/2016/01/Nala-the-cat-shark-hat.jpg'});
            }else{
              prevState.conversations.push({id: doc.id, label: res._name, userB: userB, profile: res._profilePicURL});
            }
            _this.setState(prevState);
          })
        });
      })
      .catch(function(error) {
        console.log("Error getting documents: ", error);
      });
  }

  render() {
    console.log('render!');
    console.log(this.state.conversations);
    return(
      <div className="my20">
        {this.state.conversations.map((item, idx) => {
            console.log(item);
            console.log("item!");
            firebase.firestore()
              .collection('charRoom')
              .doc(item.id)
              .collection('messages')
            return(
              <div 
              key={idx}
              // onClick={(e) => {
              //   e.stopPropagation();
              //   console.log('chat!')
              //   this.props.history.push('/chat/' + item.userB)
              // }}
              > 
                <ConversationItem
                  userB={item.userB}
                  profile={item.profile}
                  label={item.label}
                  body={item.body}
                  timestamp={item.timestamp}
                />
              </div>
            );
        })}
        {/* firebase SDK */}
        <script src="/__/firebase/6.4.0/firebase-app.js"></script>
        <script src="/__/firebase/6.4.0/firebase-auth.js"></script>
        <script src="/__/firebase/6.4.0/firebase-storage.js"></script>
        <script src="/__/firebase/6.4.0/firebase-messaging.js"></script>
        <script src="/__/firebase/6.4.0/firebase-firestore.js"></script>
        <script src="/__/firebase/6.4.0/firebase-performance.js"></script> 
      </div>
    );
  }
}

const condition = authUser => !!authUser;
export default withRouter(withAuthorization(condition)(Conversations));
