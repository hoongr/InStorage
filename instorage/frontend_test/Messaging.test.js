import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

import Messages from "../src/containers/Messaging/Messages"
import ConversationItem from "../src/containers/Messaging/ConversationItem"
import Conversations from "../src/containers/Messaging/Conversations"

describe('Messages', () => {
	const wrapper = shallow(<Messages />);

	test("renders without crashing", () => {
		expect(wrapper).toBeDefined();
	});

	test("should match the snapshot", () => {
	  expect(wrapper).toMatchSnapshot();
	});
})

describe('ConversationItem', () => {
	const wrapper = shallow(<ConversationItem />);

	test("renders without crashing", () => {
		expect(wrapper).toBeDefined();
	});

	test("should match the snapshot", () => {
	  expect(wrapper).toMatchSnapshot();
	});
})

describe('Conversations', () => {
	const wrapper = shallow(<Conversations />);

	test("renders without crashing", () => {
		expect(wrapper).toBeDefined();
	});

	test("should match the snapshot", () => {
	  expect(wrapper).toMatchSnapshot();
	});
})