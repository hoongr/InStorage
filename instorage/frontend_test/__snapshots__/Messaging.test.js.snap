// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationItem should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`Conversations should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`Messages should match the snapshot 1`] = `ShallowWrapper {}`;
